### PR TITLE
Fix format type to include strings in calculation formulas

### PR DESCRIPTION
**Pull request by zzzirvia**
## Description of the issue/feature this PR addresses

When a string variable is introduced as variable type in calculations, it fails when introducing it this same variable in the formula since the format for the variables in the formula is only for float. Here is the traceback. Error:

```
/home/silvia/senaite/zinstance/src/senaite.core/src/bika/lims/content/abstractanalysis.py:613: DeprecationWarning: BaseException.message has been deprecated as of Python 2.6
  msg = "Cannot eval formula ({}): {}".format(e.message, formula)
2024-02-08 10:51:14,136 ERROR   [senaite.core:614][waitress-0] Cannot eval formula (float argument required, not str): "ok" if %(most_1)f == "blau" else %(most_2)f
``` 

## Current behavior before PR
String not included in formula in Calculation for variables.
imatge
imatge

## Desired behavior after PR is merged
Accept string variables in formulas in calculations.
imatge

Also check works with number

imatge

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/bika/lims/content/abstractanalysis.py [hunk 1]
```diff
@@ -596,7 +596,7 @@ def calculateResult(self, override=False, cascade=False):
 
         # convert any remaining placeholders, e.g. from interims etc.
         # NOTE: we assume remaining values are all floatable!
-        formula = formula.replace("[", "%(").replace("]", ")f")
+        formula = formula.replace("[", "%(").replace("]", ")r")
```
Comment (xispa) [inline src/bika/lims/content/abstractanalysis.py]: Note that this will eventually convert any value passed in through a placeholder (e.g. `[most_1]` becomes `%(most_1)r`) using `repr()` (kind-of *official* string representation of an object). For some types, the conversion using `str` (kind-of *informal* string representation of an object) does not match with the conversion using `repr` (e.g. `datetime`).

Since we know the type of an interim beforehand (`result_type` key), I suggest to not change this line, but apply the proper converter for each interim before, at https://github.com/senaite/senaite.core/pull/2476/files#diff-85f0e866b6d6166de01fe6d6f109fe28ed787c8b4e51980a98efb2c256df70a3R557) where we do have both the keyword of the parameter/interim (e.g. `most_1` or `most_2`) and we can easily get the type with `interim.get("result_type")`. 

Basically I would expect a interim with a `result_type` value other than empty (see https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/browser/fields/interimfieldsfield.py#L101) to be converted as follows:

```python
# Convert 'Numeric' interim values using `float`. Convert the rest using `str`
converter = "s" if interim.get("result_type") else "f"
formula = formula.replace("[" + keyword + "]", "%(" + keyword + ")" + converter)
```

**Review by xispa — CHANGES_REQUESTED** [review]
Thanks @zzzirvia for your contribution. I left a comment regarding to type conversions. 

Also, would you mind to add extend the [Calculations.rst doctest](https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/tests/doctests/Calculations.rst) with your use case?

**Review by xispa — APPROVED** [review]
